src/perform_step/adams_bashforth_moulton_perform_step.jl

```diff
@@ -544,3 +544,150 @@ end
   end
   f(k, u, p, t+dt)
 end
+
+# Variable Step Size Multistep Methods
+
+function initialize!(integrator,cache::VSA3ConstantCache)
+  integrator.fsalfirst = integrator.f(integrator.uprev, integrator.p, integrator.t) # Pre-start fsal
+  integrator.kshortsize = 2
+  integrator.k = typeof(integrator.k)(integrator.kshortsize)
+
+  # Avoid undefined entries if k is an array of arrays
+  integrator.fsallast = zero(integrator.fsalfirst)
+  integrator.k[1] = integrator.fsalfirst
+  integrator.k[2] = integrator.fsallast
+end
+
+@muladd function perform_step!(integrator,cache::VSA3ConstantCache,repeat_step=false)
+  @unpack t,dt,uprev,u,f,p = integrator
+  @unpack k2,k3,grid_points,ϕstar_nm1,k,order,idx,success,tab = cache
+  
+  k1 = integrator.fsalfirst
+  cnt = integrator.iter
+
+  if cnt <= 3
+    idx = cnt
+    grid_points[idx] = t
+   elseif success
+    grid_points[1] = grid_points[2]
+    grid_points[2] = grid_points[3]
+    grid_points[3] = t
+  end
+
+  cache.grid_points .= grid_points
+
+  last_idx = k
+  next_point = t+dt
+
+  ϕ_n, ϕstar_n = ϕ_and_ϕstar!(cache, k1, next_point, last_idx)
+
+  if cnt < 4
+    cache.ϕstar_nm1 = copy(ϕstar_n)
+  else
+    cache.ϕstar_nm1 .= ϕstar_n
+  end
+
+  if cnt == 1 || cnt == 2
+    perform_step!(integrator, tab)
+    cache.k = min(k+1, order)
+    if cnt == 1
+        cache.k3 = k1
+    else
+        cache.k2 = k1
+    end
+  else
+    g = g_coefs!(cache, dt, next_point, last_idx)
+    u = uprev
+    for i = 0:k-1
+        u += dt * g[(i)+1] * ϕstar_n[(i)+1]
+    end
+
+    if integrator.opts.adaptive
+      utilde = uprev + (dt/12)*(23*k1 - 16*k2 + 5*k3) - u
+      atmp = calculate_residuals(utilde, uprev, u, integrator.opts.abstol, integrator.opts.reltol,integrator.opts.internalnorm)
+      integrator.EEst = integrator.opts.internalnorm(atmp)
+      if integrator.EEst >= 1
+        cache.success = false
```

Comment:
@YingboMa how are you doing the check on this? We should make sure it's uniform between algorithms.

Reply:
No, I didn't do this check. Maybe instead of having a boolean type for `cache.success`, we use an enum type for something like `cache.step_retcode` or `integrator.step_retcode`? Then, we can work our way to better logging.

Reply:
This is what I did for `ABDF2`
https://github.com/JuliaDiffEq/OrdinaryDiffEq.jl/blob/master/src/perform_step/bdf_perform_step.jl#L141-L143